amy/fiscal/base_views.py

```diff
@@ -31,25 +31,16 @@ def get_form_class(self):
     def get_form_kwargs(self):
         kwargs = super().get_form_kwargs()
         kwargs["queryset"] = self.get_formset_queryset(self.membership)
+        kwargs["form_kwargs"] = dict(initial={"membership": self.membership})
         return kwargs
 
     def get_context_data(self, **kwargs):
         kwargs["formset"] = self.get_form()
         return super().get_context_data(**kwargs)
 
-    def form_valid(self, form):
-        instances = form.save(commit=False)
-
-        # assign membership to any new/changed instance
-        for instance in instances:
-            instance.membership = self.membership
-            instance.save()
-
-        # remove deleted objects
-        for instance in form.deleted_objects:
-            instance.delete()
-
-        return super().form_valid(form)
+    def form_valid(self, formset):
+        formset.save()  # handles adding, updating and deleting instances
+        return super().form_valid(formset)
```

Comment:
Since we are saving before checking if it is valid, how does that affect things?  If so is there test coverage for this?

Reply:
Actually at this point we know the form is valid: https://ccbv.co.uk/projects/Django/3.1/django.views.generic.edit/FormView/#post
First `post()` checks `form.is_valid()`, and then `form_valid()` is executed.

As for the tests, `TestMembershipMembers` is a pretty big class containing tests for various cases of added/removed/mixed members.

In general, this change simplifies things, as the operations replaced (e.g. `for instance in form.deleted_objects: instance.delete()`) are covered by `formset.save(commit=True)` (commit is True by default).